File: Sources/CountriesFlagsObjC/source/Countries/UIImage+SaintKittsAndNevis.m
//
//  UIImage+SaintKittsAndNevis.m
//  CountriesFlags
//

#import "Countries/UIImage+SaintKittsAndNevis.h"

#import "Extensions/UIBezierPath+Star.h"
#import "Extensions/UIBezierPath+AffineTransform.h"

@implementation UIImage (SaintKittsAndNevis)

+ (UIImage *)flagSaintKittsAndNevisWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:151.0f/255.0f
                                           blue:57.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:1.0f
                                           green:209.0f/255.0f
                                            blue:0.0f
                                           alpha:1.0f];
    UIColor* colorBlack = [UIColor blackColor];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorRed = [UIColor colorWithRed:200.0f/255.0f
                                        green:16.0f/255.0f
                                         blue:46.0f/255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorRed setFill];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(0.0f, size.height)];
    [path addLineToPoint:CGPointMake(size.width, size.height)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path closePath];
    [path fill];

    [colorYellow setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height / 854.0f * 639.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 956.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 854.0f * 216.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 324.0f, size.height)];
    [path addLineToPoint:CGPointMake(0.0f, size.height)];
    [path closePath];
    [path fill];

    [colorBlack setFill];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height / 854.0f * 699.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 1048.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 854.0f * 154.0f)];
    [path addLineToPoint:CGPointMake(size.width / 1280.0f * 232.0f, size.height)];
    [path addLineToPoint:CGPointMake(0.0f, size.height)];
    [path closePath];
    [path fill];

    [colorWhite setFill];
    CGPoint starCenter = CGPointMake(size.width / 1280.0f * 400.0f, size.height / 854.0f * 600.0f);
    CGFloat starRadius = size.height / 854.0f * 240.0f / 2.0f;
    path = [UIBezierPath starAtPoint:starCenter
                              radius:starRadius
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:38.0f center:starCenter];
    [path fill];

    starCenter = CGPointMake(size.width / 1280.0f * 892.5f, size.height / 854.0f * 272.5f);
    path = [UIBezierPath starAtPoint:starCenter
                              radius:starRadius
                                type:FlagStarTypeFivePointed];
    [path rotateToAngle:38.0f center:starCenter];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+SaintKittsAndNevis.swift
//
//  UIImage+SaintKittsAndNevis.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Saint_Kitts_and_Nevis "wikipedia"
     
     Generates the flag of Saint Kitts and Nevis.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Saint Kitts and Nevis.
     */
    internal static func flagSaintKittsAndNevis(with size: CGSize) -> UIImage? {
        let colorGreen = UIColor(red: 0.0,
                                 green: 151.0 / 255.0,
                                 blue: 57.0 / 255.0,
                                 alpha: 1.0)
        let colorYellow = UIColor(red: 1.0,
                                  green: 209.0 / 255.0,
                                  blue: 0.0,
                                  alpha: 1.0)
        let colorBlack = UIColor.black
        let colorWhite = UIColor.white
        let colorRed = UIColor(red: 200.0 / 255.0,
                               green: 16.0 / 255.0,
                               blue: 46.0 / 255.0,
                               alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorRed.setFill()
        var path = UIBezierPath()
        path.move(to: CGPoint(x: 0.0, y: size.height))
        path.addLine(to: CGPoint(x: size.width, y: size.height))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.close()
        path.fill()

        colorYellow.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height / 854.0 * 639.0))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 956.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 854.0 * 216.0))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 324.0, y: size.height))
        path.addLine(to: CGPoint(x: 0.0, y: size.height))
        path.close()
        path.fill()

        colorBlack.setFill()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height / 854.0 * 699.0))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 1048.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 854.0 * 154.0))
        path.addLine(to: CGPoint(x: size.width / 1280.0 * 232.0, y: size.height))
        path.addLine(to: CGPoint(x: 0.0, y: size.height))
        path.close()
        path.fill()

        colorWhite.setFill()
        var starCenter = CGPoint(x: size.width / 1280.0 * 400.0, y: size.height / 854.0 * 600.0)
        let starRadius = size.height / 854.0 * 240.0 / 2.0
        path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 38.0, around: starCenter)
        path.fill()

        starCenter = CGPoint(x: size.width / 1280.0 * 892.5, y: size.height / 854.0 * 272.5)
        path = UIBezierPath.star(at: starCenter, radius: starRadius)
        path.rotate(to: 38.0, around: starCenter)
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
